### ISSUE TITLE
How to config oidc-provider to support large concurrent incoming requests?

### ISSUE THREAD

**Issue by sunpeinju** [panva/node-oidc-provider#360]
**Is your feature request related to a problem? Please describe.**
We setup a Koa server with oidc-provider, and use load test tool (artillery.io) to understand oidc-provides's performance. When testing `/token` endpoint for token refreshing, we found:
- 30 RPS (requests per second) will lead Node.js to hit 100% CPU. (Our Koa server doesn't leverage multi-core and I don't know whether oidc-provider support multi-core internally?)
- Major CPU time (70%+) is exhausted by id-token sign algorithm.
- RS256 is used to sign id-token by default, and the oidc-provider's default RSA key is 2048 bits.
- Once downgrade RSA key to 1024 bits, the server can support 90 RPS.
- If we register app with 'HS512' sign algorithm, the server can support 100 RPS.

**Describe the solution you'd like**
So I want to get suggestion for experts here:
- Does oidc-provider support to leverage multi-core CPUs internally? Or should we run Node in cluster mode and use multiple oidc-provider processes to watch same port? 
- What's the recommended id-token sign algorithm config? Shall we use RSA 256 + 2048 bit key?
- The sign/encryption functions is really CPU sensitive. Run them in major Node process significantly blocks the server's responsibility.  Does oidc-provider has any plan to optimize it? For example, run them in back end service process to leverage multi-core CPUs; replace node-forge with quicker native module?

**Describe alternatives you've considered**

**Additional context**

 - [Y] i have searched the configuration section for this feature and couldn't find it
 - [Y] i have searched the issues tracker on github for similar requests and couldn't find anything related.

**Comment by panva** [comment]
> Does oidc-provider support to leverage multi-core CPUs internally?

No. No plans. This is still just a package that you require and configure in your runtime. You're expected to deploy any cluster manager that will utilize your infrastructure cores just like if you ran your own app.

> What's the recommended id-token sign algorithm config?

You must support RS256 but your client configuration can choose a different one, there's plenty supported. Going lower than 2048bits sounds ridiculous.

> Does oidc-provider has any plan to optimize it? 

oidc-provider does not implement these functions, they all come from [node-jose](https://github.com/cisco/node-jose). `node-jose` utilizes as much native code as possible to implement these registered algorithms, where native is not possible it utilizes node-forge under the covers.

As far as digital signatures with RSASSA-PKCS1-v1_5, i have an open [node-jose PR #202](https://github.com/cisco/node-jose/pull/202) for node-jose that will make it use native node code instead of forge. Please poke that PR.

Also, [node-jose issue #203](https://github.com/cisco/node-jose/issues/203) might be something to push for.

> replace node-forge with quicker native module?

Do you have a recommendation for a module with the full range support of all [JOSE algorithms](https://tools.ietf.org/html/rfc7520)?

**Comment by sunpeinju** [comment]
@panva , thank you!
I like #202 and think using more Node's crypto module is a good idea.

I verified your PR locally, it's supper fast.
RPS raises from 30 to 200 with 2048 bit RSA256 key.

Looking forward to see it get merged.

**Comment by panva** [comment]
https://github.com/cisco/node-jose/pull/214 brings more of native node which either wasn't available or was buggy during its initial implementation into node-jose

**Comment by panva** [comment]
node-jose 1.1.0 was released, make sure to upgrade your dependency tree...

- if you're on oidc-provider ^5.x just npm upgrade oidc-provider / yarn upgrade oidc-provider
- if you're on oidc-provider ^4.x npm upgrade node-jose --depth=9999 / yarn ???
- if you're on oidc-provider < 4, upgrade to 5 :trollface:
